src/common/PresenceStatusIndicator.js

```diff
@@ -38,18 +48,21 @@ type Props = {|
  * * gray if 'offline'
  *
  * @prop [style] - Style object for additional customization.
- * @prop [presence] - UserPresence object used to determine the status from.
+ * @prop email - email of the user whose status we are showing.
  * @prop hideIfOffline - Do not render for 'offline' state.
  */
-export default class PresenceStatusIndicator extends PureComponent<Props> {
+class PresenceStatusIndicator extends PureComponent<Props> {
   render() {
-    const { presence, style, hideIfOffline } = this.props;
+    const { email, presence, style, hideIfOffline, usersByEmail, userStatus } = this.props;
 
-    if (!presence || !presence.aggregated) {
+    const userPresence = presence[email];
+    const user = usersByEmail[email];
+
+    if (!user || !userPresence || !userPresence.aggregated) {
       return null;
     }
 
-    const status = statusFromPresence(presence);
+    const status = statusFromPresenceAndUserStatus(userPresence, userStatus[user.user_id]);
```

Comment:
Cool, this nicely encapsulates the business logic (such an odd term here; "product logic"?) at work here.

A further improvement would be to also stuff the next few lines above it, with `user = usersByEmail[email]` etc., inside. I think the ideal interface for that is probably to put this all inside a selector... e.g. the `connect` call could say
```
  presenceStatus: getPresenceStatusForEmail(state, props.email),
```

That's a pattern we don't currently use much, if at all, but there are a lot of places in our code that would benefit. Detailed discussion at #3015 .

I'm fine merging with this interface as it is, though.